docs/design/2021-08-18-charsets.md

```diff
@@ -0,0 +1,270 @@
+
+# Motivation or Background
+
+Character: Various words and symbols are collectively referred to as characters. A character can be a Chinese character, an English letter, an Arabic numeral, a punctuation mark, a graphic symbol or a control symbol, etc.
+
+Charset: A collection of multiple characters is called a charset. Common character set names: ASCII character set, Unicode character set, GBK character set, etc.
+
+Character Encoding: Character encoding can be considered as a mapping rule. According to this mapping rule, a character can be mapped into other forms of data for storage and transmission in the computer. Each character set has corresponding character encoding rules, and the commonly used character set encoding rules include UTF-8 encoding, GBK encoding and so on.
+
+MySQL supports many character sets, such as `utf8mb4`, `gbk`, `gb18030`, etc. However, currently TiDB only supports 5 character sets, and there are still some problems with related character sets ([charset: incorrect encoding for `latin1` character set](https://github.com/pingcap/tidb/issues/18955)，[extended latin character set not support](https://github.com/pingcap/tidb/issues/3888)). Now it is difficult for TiDB to add a new character set. This proposal describes the implementation of the character set framework. In many traditional industries in China, `gbk`/`gb18030` is used, so this proposal uses gbk character set support as an example.
+
+## Characteristics
+
+Support gbk character set related functions:
+- Supports all code points defined by the gbk standard. Supports all code points defined by the gbk standard.
+- Support conversion with other character sets.
+  - Support data automatically converted according to the corresponding character set.
+  - Supports the use of `CAST`, `CONVERT` and other functions to convert gbk characters.
+- Support SQL statements, such as: `SET CHARACTER SET GBK`, `SET NAMES GBK`, `SHOW CHARSET`, etc.
+- It supports the comparison between strings in the gbk character set, as well as the comparison between gbk strings and strings in other character sets.
+- If the characters passed in are illegal, an error or warning is returned. If an illegal character is used in `CONVERT`, an error is returned. Otherwise, a warning will be returned.
+- Compatible with other components, including data import, migration, and so on.
+
+### Supported Collations
+
+- gbk_bin：A binary collation.
+- gbk_chinese_ci：The default collation, which supports Pinyin.
+- gbk_utf8mb4_bin: A binary collation encoded by utf8mb4.
+
+# Detailed Design
+
+After receiving the gbk character set request, this solution will convert it to the utf-8 character set, and then it will be the utf-8 character set when calculating or storing in the TiDB runtime layer and the storage layer, and finally return the result of the conversion to the gbk character set.
```

Comment:
```suggestion
After receiving the non-utf-8 character set request, this solution will convert it to the utf-8 character set, and then it will be the utf-8 character set when calculating or storing in the TiDB runtime layer and the storage layer, and finally return the result of the conversion to the original character set.
```